src/meta/expr.lean

```diff
@@ -12,11 +12,16 @@ import data.string.defs
  This file is mostly for non-tactics. Tactics should generally be placed in `tactic.core`.
 
  ## Tags
+
  expr, name, declaration, level, environment, meta, metaprogramming, tactic
 -/
 
 namespace binder_info
 
+/-!
+### Declarations about `binder_info`
```

Comment:
I don't find that comment useful. It repeats information communicated by the `namespace` declaration.

Reply:
The `namespace` declaration is not printed in doc output, and as I understand it, it would be very difficult to do that.

Reply:
By "very difficult" I mean effectively impossible. Even if we had a way to extract the begin and end lines of every namespace in a file, we wouldn't want to print all of them in the docs, and there's no way to mechanically identify the ones that are relevant.

Reply:
Does the same go with sections? If so, I think it would be worth adding this to the doc. Also, in the doc you write
```lean
/-!
### Declarations about `name`
-/
```
but I would like to argue that in this situation a one-line comment is better.

Reply:
@robertylewis Features have been added to Lean 3.5.0c to improve the extraction of documentation. Do we need something to extract the information about namespaces and sections? If so, let's put it in. I'm wary of duplicating information in comments because that leads to inconsistencies when documentation and code aren't updated properly and that makes the documentation not only useless but misleading.

@sgouezel Do you mean

```lean
/-! ### Declarations about `name` -/
```

Reply:
Yes, this is what I mean.

Reply:
> Does the same go with sections? If so, I think it would be worth adding this to the doc.

I've tried to clarify the text a bit, tell me if this is better!

> but I would like to argue that in this situation a one-line comment is better.

I've removed the new lines from single line comments.

Reply:
Yes, it's more clear like this. Thanks!

Reply:
> @robertylewis Features have been added to Lean 3.5.0c to improve the extraction of documentation. 

Yes, the current doc output (see the link in the first post that's motivating this change) makes use of the new features.

> Do we need something to extract the information about namespaces and sections? If so, let's put it in. I'm wary of duplicating information in comments because that leads to inconsistencies when documentation and code aren't updated properly and that makes the documentation not only useless but misleading.

We definitely don't want to display all uses of `namespace` and `section` in doc output. `section` is very often used to relate a handful of lemmas that have similar parameters, but are not otherwise connected in any special way. Similar for `namespace` -- often a full file is contained in a namespace, or a namespace is ended in a random place to fit a declaration with a different name. I don't think that the semantic structuring elements and the "display" structuring elements really correspond.

Reply:
Try this:

```lean
/-! foo -/

open tactic
setup_tactic_parser

@[user_command]
meta def doc_section_cmd (_ : parse $ tk "doc_section") : lean.parser unit := 
do n ← ident,
   emit_code_here sformat!"section {n} /-! ### {n} -/" .

doc_section d

run_cmd 
module_doc_strings >>= trace
-- prints: [(none, foo), (none, ### d)]
end d
```

Reply:
That would mean that, when we write `doc_section foo` instead of `section foo`, we get a corresponding heading in the documentation. We can do the same with namespaces

Reply:
This is nice but limited. I think we'll often want to display more than a single identifier as a header in the docs. Even in the `meta/expr` example, a header 

### binder_info

would look pretty weird. This would cover some simple cases (and I don't mind adding it), but I don't think it's a full replacement for sectioning comments.

Reply:
Side note: `emit_code_here` doesn't play nice with `olean_doc_strings` from @Vtec234 's PR to 3.5c. In this example, it claims the "### d" docstring is on line 1.

```lean
import tactic.core

/-! foo -/

open tactic
setup_tactic_parser

@[user_command]
meta def doc_section_cmd (_ : parse $ tk "doc_section") : lean.parser unit := 
do n ← ident,
   emit_code_here sformat!"section {n} /-! ### {n} -/" .

doc_section d

run_cmd 
olean_doc_strings >>= trace
-- prints: [(none, foo), (none, ### d)]
end d
```

Reply:
The line number can be fixed in Lean 3.5. 

If you're advocating for a default doc string like `Declarations about {foo}`, I'd prefer to have that default comment be generated in a consistent manner as in: 

```lean
emit_code_here sformat!"section {n} /-! ### Declarations about {n} -/" .
```

We can then add more details as a separate doc string underneath. If we want to depart from the default, there can be options we can set but it's ok to write out your own comment if you know where you're going with it. I just don't want a ton of comments dictated by the convention and that we write without thinking about them and that we forget (and forget to update) as soon as they're written out.

Reply:
This is a cool idea, but I would prefer that this be fixed on the doc generation side. If we are talking about changes to 3.5 to make this work, how about having namespaces and section declarations appear in the olean file?

Reply:
The point that Rob was making in this regard is that not all namespace and section headings belong in the docs. We need something to mark the difference.

Reply:
To be honest, I'm not sure I really understand the motivation for combining sectioning comments with semantic sections/namespaces.

Most uses of `section`/`namespace` should not be printed in the docs. They're there to control variable scope and declaration name, not to group a collection of similar lemmas. Even if we hold that they *should* be used in a way that can be displayed, it would be a big effort to make mathlib comply with that.

The sectioning comments in this PR, #1715, and #1716 mostly correspond to `section`s, but those `section`s aren't used semantically. They're only used to show a human reader that the contained declarations are related. With the custom to use sectioning comments, the `section`s are unnecessary.

By using a command to open a section and add a comment at the same time, you force the comment to be derived from the section name. To get docs that actually look good, you'll have to do some heuristic string processing -- remove `_`, capitalize, add some kind of template text, etc. It's not going to look natural in all cases. The alternative is a command that takes two arguments, a section name and a string for the comment, in which case... why not just keep them separate? I suspect the pattern we'd mostly see is 
```
doc_section numerals "Converting between expressions and numerals"
/!-
There are a number of ways to convert between expressions and numerals, depending on the input and
output types and whether you want to infer the necessary type classes.
-/
```
which would just as well be
```
open numerals
/!- ### Converting between expressions and numerals
...
-/
```

@cipher1024 You said you were wary of duplicating information in comments, which I get, but I'm not sure how your proposal avoids this. It's just as easy for the name of a section to become outdated as a header written in a sectioning comment.

Reply:
To be honest, I'm not sure I really see the motivation for tying sectioning comments and semantic `section`/`namespace`s together. One is purely for the human reader, the other is mainly for the system. Sometimes they'll be used together, but there are plenty of cases where you'll only need one or the other.

There are many uses of `section` and `namespace` in the library that we don't want to display in docs. In principle, we could say that they should only be used when we want them displayed, but it would be a ton of effort to get mathlib compliant with that (and I think there's no reason to).

Most of the uses of sectioning comments in this PR, #1715, and #1716 correspond to `section`s, but those `section`s don't have semantic value. They're used because it's the only standard mechanism we've used so far for displaying to the reader that a bunch of declarations are related. Sectioning comments serve the same purpose. (Comments are also used for this purpose in various places in the library, but not in any standard form.)

A command that opens a `section`/`namespace` and inserts a comment has the limitation that the name of the section has to be related to the comment. For readable docs, we'd have to do string processing -- insert the right prefix, convert `_` to space, capitalize, etc, and this still wouldn't look completely natural. More robust would be a command that takes two arguments, the namespace name and the text for the comment. But then why not just keep the two separate? I think the most common pattern would be 

```
doc_section numerals "Converting between expressions and numerals"
/-!
There are a number of ways to convert between expressions and numerals, depending on the input and
output types and whether you want to infer the necessary type classes.
-/
```
which doesn't seem any better than
```
section numerals
/-!
### Converting between expressions and numerals
There are a number of ways to convert between expressions and numerals, depending on the input and
output types and whether you want to infer the necessary type classes.
See also the tactics `expr.of_nat`, `expr.of_int`, `expr.of_rat`.
-/
```

@cipher1024 You said you're wary of duplicating information because it can become out of date, which is a reasonable concern, but I'm not sure how your proposal fixes that. It's just as easy for the name of a section to become outdated as it is for a comment.

Reply:
> `emit_code_here` doesn't play nice with `olean_doc_strings`
Internally it uses `with_input`, which seems to replace the scanner state with a fresh one initialized at line 0, column 0. It could be initialized with the current position of the real scanner, but that might break other things - I would have to take a close look at it.

Reply:
@robertylewis You posted the same comment twice. Can you delete one please? I'm allowed to do it but I want to make sure that, if there's a better version, you can choose it yourself.

Reply:
Oh, weird, sorry. GitHub gave me a bizarre error message when I tried to post the first one and it didn't show up for a bit so I rewrote it. Done, I think they said pretty much the same thing.

Reply:
From the PR:

```quote
It is common to structure a file in sections, where each section contains related declarations.
By describing the sections with module documentation `/-! ... -/` at the beginning,
these sections can be seen in the documentation.
```

I understand that you're (@robertylewis) are advocating matching some `section` and `namespace` with a heading in a module doc string. My point is that if you're supporting this practice, you're advocating the widespread verbatim replication of information in comments like so:

```lean
section foo_bar
/-! ### Declarations about `foo_bar` -/
...
end foo_bar
```

When renaming a declaration called `foo_bar` to `new_foo_bar`, the parser and type checker will force us to update code consistently but not comments and we can end up with:

```lean
section new_foo_bar
/-! ### Declarations about `foo_bar` -/
...
end new_foo_bar
```

If `foo_bar` is now the name of something else, it favors confusion. What I'm suggesting is the following:

```lean
doc_section foo_bar
...
end foo_bar
```

because updating the section name cannot be done without updating the heading. This approach is made possible because ``Declarations about `foo_bar` `` has no other technical information than the name `foo_bar`. If ``Declarations about `foo_bar` `` is the convention that you want to use in the documentation, although departing from the convention is harder with the `doc_section` command, it will avoid the proliferation of comments that do nothing but repeat technical information already present in the code for the mere sake of the documentation.

In sum, I do not purport to avoid typos or mistakes in the name of sections and namespaces but only to make a section name and its accompanying header comment easier to keep synched up.

Reply:
> I understand that you're (@robertylewis) are advocating matching some `section` and `namespace` with a heading in a module doc string.

I can adjust the text there to clarify. There are many many places in mathlib where we have clusters of related declarations. Sometimes they're in a `section`; sometimes they're in a `namespace`; sometimes there's a `--` comment describing the "following declarations" or similar; often there's no explicit sectioning at all.

> When renaming a declaration called `foo_bar` to `new_foo_bar`, the parser and type checker will force us to update code consistently

Only if it's a `namespace`. `section` names have no meaning. Every use of this in #1715 and #1716 is a section.

Reply:
Maybe using `meta/expr.lean` was a bad example. Look at `tactic/core.lean`. Right at the beginning, I see a cluster of declarations manipulating numerals:
```### Tactics for creating numeral expressions```

A bit later, I see a cluster in the `lean.parser` namespace related to `emit_code_here`. I wouldn't want to display `### Declarations about lean.parser` in the docs, the namespace isn't really the relevant point.

Later I see sections about manipulating declarations, creating and destructing binder expressions, etc. They all belong in the tactic namespace; it doesn't matter whether they're in a `section` or not; the most descriptive headers for the sections don't all fit into the "Declarations about _" pattern. I just don't see what you gain by using a command to match a section name and a doc string here, and I see that you lose legibility of the doc string.

Reply:
That said: I don't particularly *mind* adding the command, for the situation where you want to (1) open a semantic `section`/`namespace` and (2) add a one line sectioning comment that's formatted in a very specific way. I think its use is pretty limited, but as long as it makes some effort to make the section name displayable, it doesn't hurt.